File: uniconn-node/src/conn.ts
/**
 * uniconn — Universal Connection Interface for Node.js
 *
 * This module defines the core IConn, IListener, and IDialer interfaces
 * that abstract various network protocols (TCP, WebSocket, QUIC,
 * WebTransport, KCP) behind a unified, net.Conn-inspired API.
 */

/** A network address descriptor. */
export interface Addr {
  /** Protocol name: "tcp", "websocket", "quic", "webtransport", "kcp" */
  readonly network: string;
  /** Address string: "127.0.0.1:8080", "wss://example.com/ws", etc. */
  readonly address: string;
}

/**
 * IConn — Universal connection interface, modeled after Go's net.Conn.
 *
 * Each protocol adapter implements this interface so that application
 * code can read, write, and close connections identically regardless
 * of the underlying transport.
 */
export interface IConn {
  /**
   * Read data into the provided buffer.
   * Resolves with the number of bytes read.
   * Resolves with 0 at EOF (peer closed the connection).
   */
  read(buffer: Uint8Array): Promise<number>;

  /**
   * Write data to the connection.
   * Resolves with the number of bytes written.
   */
  write(data: Uint8Array): Promise<number>;

  /** Close the connection. */
  close(): Promise<void>;

  /** Returns the local address. */
  localAddr(): Addr;

  /** Returns the remote address. */
  remoteAddr(): Addr;

  /**
   * Set both read and write timeout in milliseconds.
   * 0 means no timeout.
   */
  setDeadline(ms: number): void;

  /** Set read timeout in milliseconds. 0 means no timeout. */
  setReadDeadline(ms: number): void;

  /** Set write timeout in milliseconds. 0 means no timeout. */
  setWriteDeadline(ms: number): void;
}

/** IListener — Server-side listener interface. */
export interface IListener {
  /** Accept and return the next incoming connection. */
  accept(): Promise<IConn>;

  /** Stop listening and close the listener. */
  close(): Promise<void>;

  /** Returns the address the listener is bound to. */
  addr(): Addr;
}

/** Options for dialing a connection. */
export interface DialOptions {
  /** An AbortSignal to cancel the dial operation. */
  signal?: AbortSignal;
  /** Dial timeout in milliseconds. */
  timeout?: number;
}

/** IDialer — Client-side dialer interface. */
export interface IDialer {
  /** Establish a connection to the given address. */
  dial(address: string, options?: DialOptions): Promise<IConn>;
}

File: uniconn-py/uniconn/conn.py
"""
Core connection interfaces for uniconn.

All protocol adapters implement these interfaces so that application code
can work with any transport transparently.
"""

from __future__ import annotations

import abc
from dataclasses import dataclass


@dataclass(frozen=True)
class Addr:
    """Network address descriptor."""
    network: str  # "tcp", "websocket", "kcp", etc.
    address: str


class Conn(abc.ABC):
    """
    Universal connection interface, modeled after Go's net.Conn.

    Each protocol adapter implements this so that application code can
    read, write, and close connections identically regardless of transport.
    """

    @abc.abstractmethod
    async def read(self, buf: bytearray) -> int:
        """Read data into buffer. Returns number of bytes read. 0 = EOF."""
        ...

    @abc.abstractmethod
    async def write(self, data: bytes) -> int:
        """Write data. Returns number of bytes written."""
        ...

    @abc.abstractmethod
    async def close(self) -> None:
        """Close the connection."""
        ...

    @abc.abstractmethod
    def local_addr(self) -> Addr | None:
        """Return local address, if available."""
        ...

    @abc.abstractmethod
    def remote_addr(self) -> Addr | None:
        """Return remote address, if available."""
        ...


class Listener(abc.ABC):
    """Server-side listener interface."""

    @abc.abstractmethod
    async def accept(self) -> Conn:
        """Accept and return the next incoming connection."""
        ...

    @abc.abstractmethod
    async def close(self) -> None:
        """Stop listening."""
        ...

    @abc.abstractmethod
    def addr(self) -> Addr:
        """Return bound address."""
        ...


class Dialer(abc.ABC):
    """Client-side dialer interface."""

    @abc.abstractmethod
    async def dial(self, address: str) -> Conn:
        """Establish a connection to the given address."""
        ...
